swift/ql/lib/codeql/swift/regex/Regex.qll

```diff
@@ -341,17 +327,50 @@ abstract class RegexEval extends CallExpr {
       any(RegexAdditionalFlowStep s).setsParseMode(setNode, result, true) and
       // reaches this eval
       (
-        RegexParseModeFlow::flow(setNode, this.getRegexInput()) or
-        RegexParseModeFlow::flow(setNode, this.getAnOptionsInput())
+        RegexParseModeFlow::flow(setNode, this.(PotentialRegexEval).getRegexInput()) or
+        RegexParseModeFlow::flow(setNode, this.(PotentialRegexEval).getAnOptionsInput())
       )
     )
   }
 }
 
+/**
+ * A call that may evaluate a regular expression. Extend this abstract class to
+ * add new regular expression evaluation models.
+ */
+abstract class PotentialRegexEval extends CallExpr {
+  /**
+   * Gets the input to this call that is the regular expression being evaluated.
+   * This may be a regular expression object or a string literal.
+   */
+  abstract DataFlow::Node getRegexInput();
+
+  /**
+   * Gets the input to this call that is the string the regular expression is evaluated on.
+   */
+  abstract DataFlow::Node getStringInput();
+
+  /**
+   * Gets a dataflow node for the options input that might contain parse mode
+   * flags (if any).
+   */
+  DataFlow::Node getAnOptionsInput() { none() }
+
+  /**
+   * Holds if this is an actual regular expression evalaution. If this does not
+   * hold, the potential regular expression evaluation should be discarded.
+   *
+   * This mechanism exists so that we have something to track flow of options
+   * into to (for example an `NSString.CompareOptions.regularExpression` option)
+   * before deciding whether a regular expression is actually evaluated.
+   */
+  predicate doesEvaluate() { any() }
```

Comment:
So with this we can actually be in a situation where a call is an `PotentialRegexEval`, but `doesEvaluate` doesn't hold so it's not _actually_ a regex evaluator? That seems like a weird design to me, I think 🤔. Why not just stick the code from this predicate into the charpred? That is, if I have code such as:
```ql
class MyEvaluator extends PotentialRegexEval {
  MyEvaluator() { ... }
  
  override predicate doesEvaluate() { /* check if it evaluates a regular expression */ }
}
```
why not rewrite this as:
```ql
class MyEvaluator extends PotentialRegexEval {
  MyEvaluator() { ... and /* check if it evaluates a regular expression */ }
}
```
and then there'd be no need for `doesEvaluate`?

Reply:
I've just pushed an update, which feels like a big improvement.  Thanks for pushing me on this.

I did have to move the new data flow config into `Regex.qll` rather than `RegexTracking.qll` so it can see the now private `...PotentialRegexEval` class - which doesn't feel natural right now.  But I plan on splitting the models out into their own file anyway, at which point these two being together will make sense and they'll be out of the way of the core regex logic.

Let me know if you have any further thoughts!

Reply:
Awesome. I think this is a much better design. Thanks for looking into this!